sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/v2/SchemaOnWriteDataSource.scala

```diff
@@ -0,0 +1,26 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.spark.sql.execution.datasources.v2
+
+import org.apache.spark.sql.connector.catalog.TableProvider
+import org.apache.spark.sql.sources.DataSourceRegister
+
+/**
+ * A data source that is given the DataFrame's schema upon write operations.
+ */
+trait SchemaOnWriteDataSource extends TableProvider with DataSourceRegister {
```

Comment:
One option is to add this trait, and another option is to officially make `FileDataSourceV2` a public developer API. Does this "schema-on-write" behavior make sense for non-file data sources?

cc @gengliangwang

Reply:
+1, making `FileDataSourceV2` a public developer API seems simpler